src/main/java/org/threadly/concurrent/NoThreadScheduler.java

```diff
@@ -240,57 +231,70 @@ private int tick(ExceptionHandler exceptionHandler, boolean resetCancelTickIfNoT
   public int blockingTick(ExceptionHandler exceptionHandler) throws InterruptedException {
     int initialTickResult = tick(exceptionHandler, false);
     if (initialTickResult == 0) {
-      currentlyBlocking = true;
+      Thread currentThread = Thread.currentThread();
+      // we already tried to optimistically run something above, so we now must prepare to park
+      // and park if we still find nothing to execute
+      if (! blockingThread.compareAndSet(null, currentThread)) {
+        throw new IllegalStateException("Another thread is already blocking!!");
+      }
       try {
-        synchronized (taskNotifyLock) {
-          while (true) {
-            /* we must check the cancelTick once we have the lock 
-             * since that is when the .notify() would happen.
-             */
-            if (tickCanceled) {
-              tickCanceled = false;
-              return 0;
+        while (true) {
+          /* we must check the cancelTick once we have the lock 
+           * since that is when the .notify() would happen.
+           */
+          if (tickCanceled) {
+            tickCanceled = false;
+            return 0;
+          } else if (currentThread.isInterrupted()) {
+            throw new InterruptedException();
+          }
+          TaskWrapper nextTask = getNextTask(highPriorityQueueSet, lowPriorityQueueSet, 
+                                             maxWaitForLowPriorityInMs);
+          if (nextTask == null) {
+            TaskWrapper nextStarvableTask = starvablePriorityQueueSet.getNextTask();
+            if (nextStarvableTask == null) {
+              LockSupport.park();
+            } else {
+              long nextStarvableTaskDelay = nextStarvableTask.getScheduleDelay();
+              if (nextStarvableTaskDelay > 0) {
+                LockSupport.parkNanos(Clock.NANOS_IN_MILLISECOND * nextStarvableTaskDelay);
+              } else {
+                // starvable task is ready to run, so break loop
+                break;
+              }
             }
-            TaskWrapper nextTask = getNextTask(highPriorityQueueSet, lowPriorityQueueSet, 
-                                               maxWaitForLowPriorityInMs);
-            if (nextTask == null) {
+          } else {
```

Comment:
the park logic is still different from PriorityScheduler because of how we need to queue the threads to be unparked.  I don't think we could gain much there unless you are seeing something I am not?